QQuickPaintedItem support

devmil opened the pull request:
This PR adds QQuickPaintedItem support to Qml.Net.

I use this currently in a QML.Net based terminal application, so it seems to work.

The approach is:
1. Qml.Net provides a new registered type: QuickItems
2. Qml.Net provides a .Net base class for painted items: QmlNetPaintedItem
3. Users create a subclass of QmlNetPaintedItem on .Net side and implement the paint() method
4. The C++ side makes sure that the paint() call gets routed to the .Net class that can use the QPainter that is passed to do the actual drawing
5. Qml.Net makes sure that all properties, methods and signals that the QmlNetPaintedItem derived type has are mirrored to QML just as they are with any other type that gets registeres with QML

![QmlNetInnerWorkingsPaintedItem](https://user-images.githubusercontent.com/6693130/95000356-5dc54080-05c0-11eb-9d2d-ebe24157b580.png)

Example usage:
```QML
import Qml.Net 1.0
// [...]
    TerminalQuickPaintedItem {
        id: terminalContent
        anchors.fill: parent
        anchors.margins: terminalMargins
// [...]
    }
``` 
```C#
public class TerminalQuickPaintedItem : QmlNetQuickPaintedItem
{
    public TerminalViewModel TerminalViewModel { get; set; }

    public override void Paint(NetQPainter painter)
    {
        //implement paint logic
    }
}

//[...]

Qml.Net.Qml.RegisterPaintedQuickItemType<TerminalQuickPaintedItem>(@namespace, versionMajor, versionMinor);
```
Solves #198

pauldotknopf committed:
Merge branch 'hotfix/0.10.1'

pauldotknopf committed:
Merge branch 'release/0.11.0'

devmil committed:
First support for QQuickPaintedItem

Implements some basic drawing functionalities

devmil committed:
Less string passing during drawing

Colors get registered and are referenced via id
Font parameters can be set separately

devmil committed:
Color string to id conversion is now an implementation detail of QmlNetPaintedItem

devmil committed:
Adds Font metrics functionality

devmil committed:
Adds DrawText overload with bounding rectangle

devmil committed:
Dont copy the actions on each draw but lock the vector

devmil committed:
Optional Text support for QmlNetPaintedItem (including preedit)

devmil committed:
Adds Flags to the DrawText (Rect) method

pauldotknopf commented:
Sorry for the delay, just getting to this.

I like this approach, and it could work well. The only thing I'm not a fan of is the ```onXChanged``` methods, and communicating that to the .NET paint handler. I'd prefer an approach that allows you to inherit a .NET ```QmlNetQuickPaintedItem``` that has methods that can be overridden, just as you would do with C++. We'd register this derived ```QmlNetQuickPaintedItem``` with ```QmlNet.RegisterPaintedType<T>(...) where T: QmlNetQuickPaintedItem```.

Then, we'd have something like this:

```json
MyCustomPaintedType {
}
```
```csharp
public class MyCustomPaintedType : QmlNetQuickPaintedItem
{
    public override void Paint(QPainter painter) { ... }
    public override void OnWidthChanged(int width) { ... }
}
```

This is me thinking out-loud though. You are currently in the trenches, so I'd really like to hear what you'd consider the pros/cons of each approach in comparison.

devmil committed:
Adds an interface to interact with a QPainter

This interface can be implemented by a QPainter abstraction in the future
For now QmlNetPaintedItem implements INetQPainter

devmil committed:
Merge remote-tracking branch 'upstream/develop' into feature/QQuickPaintedItem_support

devmil committed:
Native submodule: Adds QMutex include

devmil committed:
Native submodule: Adds missing include

devmil committed:
Adds a lot of the QPainter interface

devmil committed:
Fixes Polygon point handling

Adds new QPainter API to QmlNetPaintedItem

devmil committed:
Renames QmlNetPaintedItem to QmlNetRecordingPaintedItem

devmil committed:
Deactivates painthandler reset in destructor of RecordingPaintedItem

devmil committed:
Registering a QuickPaintedItem

devmil committed:
Hide QmlNetQuickPaintedItem from QML

devmil committed:
First theoretically working PaintedItem implementation

devmil committed:
getStringSize is now static

devmil committed:
new native submodule version

devmil commented:
Ok, seems like you were right :) I implemented the QuickPaintedItem in a way that the C# class implements Paint and controls the painting directly from .Net and integrated it in my Terminal app I'm working on.
Result: as fast as before.

It was a little bit tricky to have the NetReference managed .Net object that mimics a QuickPaintedItem to be a real QuickItem derived type in C++ but I think it's working now. I basically had to tweak the metaobject's superclass to point to the QuickItem types metaobject and prvent the .Net part from collecting QmlNetQuickPaintedItem's data.

The way it now works:
*1. derive a class from QmlNetQuickPaintedItem*
*2. Register that new class using the new registration api extension*
*3. make sure that the data is accessible from the painted item*
*4. implement paint()*

```
public class TerminalQuickPaintedItem : QmlNetQuickPaintedItem
{
    public TerminalViewModel TerminalViewModel { get; set; }

    public override void Paint(INetQPainter painter)
    {
        //implement paint logic
    }
}

//[...]

Qml.Net.Qml.RegisterPaintedQuickItemType<TerminalQuickPaintedItem>(@namespace, versionMajor, versionMinor);

//[...]

    TerminalQuickPaintedItem {
        id: terminalContent
        anchors.fill: parent
        anchors.margins: terminalMargins
//[...]
    }
```

I want to extend the tests and I'm not sure what properties of a QQuickPaintedItem might be relevant on the .Net side. I included height and width for now.

devmil committed:
new native submodule (build error fix)

devmil committed:
Cleanup

Removes RecordingPaintedItem
Removes INetQPainter (functionality moved to NetQPainter)

devmil committed:
references native that fixes type registration

Blquinn commented:
Great progress, looks awesome!

Instead of having drawImage with a byte[], would it be worth having

drawImage(System.Drawing.Image img);
and
drawBitmap(System.Drawing.Bitmap bmp);

variants? They could just do the conversion to byte[] in dotnet land.

devmil commented:
> Great progress, looks awesome!
> 
> Instead of having drawImage with a byte[], would it be worth having
> 
> drawImage(System.Drawing.Image img);
> and
> drawBitmap(System.Drawing.Bitmap bmp);
> 
> variants? They could just do the conversion to byte[] in dotnet land.

Thanks for the input. Not sure if we want to introduce a dependency to System.Drawing.Common for this purpose?
But thinking about that method we should at least add a variant that supports passing the file path instead of its data

devmil committed:
Draw Image file

devmil committed:
Adds testing infrastructure for QmlNetQuickPaintedItem

devmil committed:
DrawArc test

devmil committed:
DrawConvexPolygon test

pauldotknopf commented on src/net/Qml.Net/QmlNetQuickPaintedItem.cs:
```diff
@@ -0,0 +1,194 @@
+        internal QmlNetPaintedItemCallbacksImpl()
+        {
+            _setRef = SetRef;
+            GCHandle.Alloc(_setRef);
```
Are these freed?

devmil replied:
nope, good spot. I assume we have the same problem in Callbacks.cs then? (This is where I got the inspiration for that Callback implementation)

pauldotknopf replied:
```Callbacks.cs``` doesn't have the issue, because it's static/singleton.

devmil replied:
As is `QmlNetPaintedItemCallbacksImpl`.
Btw: not storing the return value of GCHandle.Alloc is immediately freeing the handle? (Asking because of Callback.cs)

pauldotknopf reviewed (commented):
What is the purpose of ```supportsTextInput```?

devmil commented:
> What is the purpose of `supportsTextInput`?

This enables InputMethod support which allows to take part in e.g. preedit text sessions. This is needed when the PaintedItem wants to receive text.

devmil committed:
Fixes GCHandle handling in QmlNetQuickPaintedItem

devmil committed:
Adds setWindowIcon support to QGuiApplication

devmil committed:
Adds DrawEllipse and DrawLine test

devmil committed:
Adds DrawPoint test

devmil committed:
Adds DrawPolygon test

devmil committed:
Adds DrawRoundedRect test

devmil committed:
Draw Pie Unit Test

devmil committed:
Test source formatting

devmil committed:
DrawPolyline Unit test

devmil committed:
Erase Rect Unit Test

devmil committed:
Unit Tests for ClipRect and Opacity

devmil committed:
Disables transformations

Real world use cases would want to use the QTransform functionality instead of passing the raw transformation matrix

devmil committed:
Unit Test for Sheer

devmil committed:
Adds Unit test for Translate

devmil commented:
Drive-by fixes #181 I assume.
I think it makes sense to wait for PR #215 as it also contains changes in the native part?

devmil committed:
Cleanup

devmil committed:
Merge remote-tracking branch 'upstream/develop' into feature/QQuickPaintedItem_support

pauldotknopf commented:
I'm not sure what's going on with #215. I eventually had time to review it (took me a while, I know, my fault), but I'm not sure the PR is going anywhere. I moved the native module ```master``` branch to where it was previously.

Anyways, this PR looks good to me.

pauldotknopf commented:
After the conflicts are cleared, I think we can merge it.

devmil committed:
Link to native/origin/master

devmil committed:
Merge remote-tracking branch 'upstream/develop' into feature/QQuickPaintedItem_support

devmil committed:
Re-Trigger CI

devmil commented:
Hi @pauldotknopf will you merge this PR when you think it makes sense or shall I?

pauldotknopf commented:
It looks like there are some conflicts? Can you take care of them? After that, feel free to merge.

devmil commented:
@pauldotknopf what conflicts do you mean?

pauldotknopf commented:
![image](https://user-images.githubusercontent.com/523150/98128618-1d3e3880-1e86-11eb-90eb-a459329e94cd.png)

vadi2 commented:
Try 'squash and merge' option?

devmil commented:
Seems like your Github UI wants to rebase?
![17DFBDFC-E03C-45BD-9FEB-A78BCBB3E903_1_105_c](https://user-images.githubusercontent.com/6693130/98138605-927c2e80-1ec3-11eb-8098-d0284131bfe3.jpeg)
